### PR TITLE
feat(npu): Support returning indexer top-k results

### PR DESCRIPTION
**Pull request by floatlibai**
## Motivation

`enable_return_indexer_topk` currently supports the CUDA indexer path, but the NPU backend does not create an indexer-topk capturer or expose the top-k results produced by NPU attention indexers.

  Ascend NPU already provides indexer implementations for:

  - DeepSeek DSA models, through `torch_npu.npu_lightning_indexer`
  - DeepSeek-V4 C4 indexers, through `torch.ops.custom.npu_quant_lightning_indexer`

  However, their results were not connected to the existing request-level return pipeline. As a result, NPU inference requests could not return indexer top-k metadata even when `--enable-return-indexer-topk` and `return_indexer_topk=true` were specified.

  This PR connects the NPU producer paths to the existing indexer-topk capturer and response pipeline while preserving the same output format and layer-index semantics as the CUDA implementation.

## Modifications

- Enable indexer-topk capturer creation on both CUDA and NPU backends.
    - Model-specific layer counts continue to be resolved through `get_num_indexer_layers()`.
    - DSA models use one capturer slot per transformer layer.
    - DeepSeek-V4 uses one capturer slot per C4 compression layer.

- Add indexer-topk capture to the NPU DSA path.
    - Capture results from `forward_dsa_prepare_npu()`.
    - Cover both freshly computed top-k results and `skip_topk` layers that reuse the previous layer's result.
    - Preserve the transformer layer id in the capturer layout.

- Add indexer-topk capture to the NPU DeepSeek-V4 C4 path.
    - Capture the output of `npu_quant_lightning_indexer`.
    - Use `token_to_kv_pool.layer_mapping[].compress_layer_id` to map raw transformer layers to contiguous C4 capturer slots.
    - Keep the captured indices in the logical compressed-token coordinate system used by the NPU sparse attention path.

- Extend overlap result handling to NPU.
    - Reuse the existing `TopkCaptureOutput` and result-finalization pipeline.
    - Add NPU support to asynchronous device-to-host copies through `torch.get_device_module()`.
    - Preserve the existing synchronous path when overlap scheduling is disabled.

- Add unit tests covering:
    - NPU DSA capturer initialization.
    - DeepSeek-V4 C4 layer-count resolution.
    - Non-indexer models being disabled through a zero indexer-layer count.
    - Invalid `index_topk` configurations.

## Validation

test plan:
  - Added CPU unit coverage for NPU capturer selection and layer-count handling.
  - NPU end-to-end shape/range validation tested on Ascend hardware.

shape/range test result (modified from `test/registered/8-gpu-models/test_return_indexer_topk.py`):
- GLM5.2 bf16 - 16 layers
```sh
> python test_indexer_replay_en.py --base-url http://127.0.0.1:8000 --num-layers 16 --topk-size 2048

Target server: http://127.0.0.1:8000
Number of test texts: 2
Expected num_indexer_layers=16, index_topk=2048,
Sending requests...

--- Text: 'What is the capital of France?'  Shape: (22, 16, 2048) ---
  [PASS] shape/range check passed

--- Text: 'Solve: 2 + 3 = ?'  Shape: (25, 16, 2048) ---
  [PASS] shape/range check passed

============================================================
Result: All checks passed ✅
```

- GLM5.2 w8a8 - full layers
```sh
> python test_indexer_replay_en.py --base-url http://141.61.54.248:6677 --num-layers 78 --topk-size 2048

Target server: http://141.61.54.248:6677
Number of test texts: 2
Expected num_indexer_layers=78, index_topk=2048,
Sending requests...

--- Text: 'What is the capital of France?'  Shape: (22, 78, 2048) ---
  [PASS] shape/range check passed

--- Text: 'Solve: 2 + 3 = ?'  Shape: (25, 78, 2048) ---
  [PASS] shape/range check passed

============================================================
Result: All checks passed ✅
```

- DeepSeek v4 flash bf16 - full layers
```sh
> python test_indexer_replay_en.py --base-url http://0.0.0.0:8001 --num-layers 21 --topk-size 512

Target server: http://0.0.0.0:8001
Number of test texts: 2
Expected num_indexer_layers=21, index_topk=512,
Sending requests...

--- Text: 'What is the capital of France?'  Shape: (22, 21, 512) ---
  [PASS] shape/range check passed

--- Text: 'Solve: 2 + 3 = ?'  Shape: (24, 21, 512) ---
  [PASS] shape/range check passed

============================================================
Result: All checks passed ✅
```
e2e testcase result:
- deepseek v4 flash bf16 (reduce to 8 layers)
```sh
 >  pytest test/registered/npu/basic_function/rl/indexer_replay/test_npu_return_indexer_topk_deepseek_v4.py
=================================================================== test session starts ===================================================================
platform linux -- Python 3.12.13, pytest-8.3.2, pluggy-1.6.0
rootdir: /sgl-workspace/sglang/test
configfile: pytest.ini
plugins: xdist-3.6.1, anyio-4.15.1
collected 1 item

test/registered/npu/basic_function/rl/indexer_replay/test_npu_return_indexer_topk_deepseek_v4.py .                                                  [100%]

==================================================================== warnings summary =====================================================================
../../usr/local/python3.12.13/lib/python3.12/site-packages/_pytest/config/__init__.py:1437
  /usr/local/python3.12.13/lib/python3.12/site-packages/_pytest/config/__init__.py:1437: PytestConfigWarning: Unknown config option: asyncio_mode

    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

../../usr/local/python3.12.13/lib/python3.12/site-packages/torch/jit/_script.py:362: 14 warnings
  /usr/local/python3.12.13/lib/python3.12/site-packages/torch/jit/_script.py:362: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= 1 passed, 15 warnings in 103.62s (0:01:43) ========================================================
```

- glm5.2 bf16 (reduce to 6 layers)
```sh
> pytest test/registered/npu/basic_function/rl/indexer_replay/test_npu_return_indexer_topk_glm52.py
=================================================================== test session starts ===================================================================
platform linux -- Python 3.12.13, pytest-8.3.2, pluggy-1.6.0
rootdir: /sgl-workspace/sglang/test
configfile: pytest.ini
plugins: xdist-3.6.1, anyio-4.15.1
collected 1 item

test/registered/npu/basic_function/rl/indexer_replay/test_npu_return_indexer_topk_glm52.py .                                                        [100%]

==================================================================== warnings summary =====================================================================
../../usr/local/python3.12.13/lib/python3.12/site-packages/_pytest/config/__init__.py:1437
  /usr/local/python3.12.13/lib/python3.12/site-packages/_pytest/config/__init__.py:1437: PytestConfigWarning: Unknown config option: asyncio_mode

    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

../../usr/local/python3.12.13/lib/python3.12/site-packages/torch/jit/_script.py:362: 14 warnings
  /usr/local/python3.12.13/lib/python3.12/site-packages/torch/jit/_script.py:362: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= 1 passed, 15 warnings in 123.55s (0:02:03) ========================================================
```
cpu testcase result:
```sh
> pytest ../sglang/test/registered/unit/state_capturer/test_indexer_topk_npu.py
pytest test/registered/unit/state_capturer/test_indexer_topk_npu.py
=================================================================== test session starts ===================================================================
platform linux -- Python 3.12.13, pytest-8.3.2, pluggy-1.6.0
rootdir: /sgl-workspace/sglang/test
configfile: pytest.ini
plugins: xdist-3.6.1, anyio-4.15.1
collected 4 items

test/registered/unit/state_capturer/test_indexer_topk_npu.py ....                                                                                   [100%]

==================================================================== warnings summary =====================================================================
../../usr/local/python3.12.13/lib/python3.12/site-packages/_pytest/config/__init__.py:1437
  /usr/local/python3.12.13/lib/python3.12/site-packages/_pytest/config/__init__.py:1437: PytestConfigWarning: Unknown config option: asyncio_mode

    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

../../usr/local/python3.12.13/lib/python3.12/site-packages/torch/jit/_script.py:362: 14 warnings
  /usr/local/python3.12.13/lib/python3.12/site-packages/torch/jit/_script.py:362: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================= 4 passed, 15 warnings in 11.45s =============================================================
```
## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35730322528](https://github.com/sgl-project/sglang/actions/runs/35730322528)
Latest PR Test (Extra): :x: [Run #35730322442](https://github.com/sgl-project/sglang/actions/runs/35730322442)
Latest PR Test (AMD ROCm 10): :x: [Run #35730322935](https://github.com/sgl-project/sglang/actions/runs/35730322935)

### CODE REVIEW

File: python/sglang/srt/state_capturer/indexer_topk.py [hunk 1]
```diff
@@ -133,6 +132,9 @@ def _create_indexer_capturer_raw(
     if num_indexer_layers == 0:
         logger.warning("No indexer layers found, IndexerTopkCapturer disabled")
         return None
+    if index_topk == 0:
+        logger.warning("Invalid index_topk = 0, IndexerTopkCapturer disabled")
+        return None
```
Comment (Ronald1995) [inline python/sglang/srt/state_capturer/indexer_topk.py:137]: it's better to guard it like

```python

if index_topk is None or index_topk <= 0:
    logger.warning(
        "Invalid index_topk=%s, "
        "IndexerTopkCapturer disabled",
        index_topk,
    )
    return None
```

File: python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py [hunk 2]
```diff
@@ -889,6 +890,10 @@ def forward_c4_indexer(
             q, weights = self._forward_prepare(c4_indexer, x, q_lora, forward_batch)
         topk_idxs = self._forward_indexer(c4_indexer, x, q, weights, forward_batch)
```
Comment (Ronald1995) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py:891]: please show your e2e validation result and add e2e cases for these two models.

File: python/sglang/srt/state_capturer/indexer_topk.py (hunk 1 above)
Comment (floatlibai) [reply python/sglang/srt/state_capturer/indexer_topk.py:139]: I have changed this to `<=0`

File: test/registered/npu/basic_function/rl/indexer_replay/test_npu_return_indexer_topk_deepseek_v4.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+NUM_INDEXER_LAYERS = 3
+INDEX_TOPK = 512
+
+
+class TestNPUDeepSeekV4ReturnIndexerTopk(CustomTestCase):
+    """End-to-end validation of DeepSeek-V4 C4 NPU indexer-topk responses."""
+
+    @classmethod
+    def setUpClass(cls):
+        cls.process = popen_launch_server(
```
Comment (floatlibai) [inline test/registered/npu/basic_function/rl/indexer_replay/test_npu_return_indexer_topk_deepseek_v4.py:56]: lack env setup
```sh
# skip gpu branch
export SGLANG_OPT_FP8_WO_A_GEMM=0
export SGLANG_OPT_USE_OVERLAP_STORE_CACHE=False
export SGLANG_DSV4_FP4_EXPERTS=False
export SGLANG_OPT_FUSE_WQA_WKV=0
export SGLANG_OPT_BF16_FP32_GEMM_ALGO=torch
export SGLANG_OPT_USE_FUSED_HASH_TOPK=False
export SGLANG_OPT_USE_TILELANG_MHC_PRE=False
export SGLANG_OPT_DEEPGEMM_HC_PRENORM=False
export SGLANG_OPT_USE_TILELANG_MHC_POST=False

export SGLANG_ENABLE_WAR_BARRIER=1
export SGLANG_FORCE_COARSE_WAR_BARRIER=1


python3 -m sglang.launch_server --model-path ${MODEL_PATH} \
    --page-size 128 \
    --tp-size 4 \
    --trust-remote-code \
    --attention-backend dsv4 \
    --device npu \
    --host 0.0.0.0 --port 8001 \
    --mem-fraction-static 0.6 \
    --kv-cache-dtype auto \
    --enable-dp-lm-head \
    --disable-cuda-graph \
    --enable-dp-attention --dp-size 4 \
    --enable-return-indexer-topk \
    --json-model-override-args '{"num_hidden_layers": 8, "compress_ratios": [0, 0, 4, 128, 4, 128, 4, 128]}'
```

File: test/registered/npu/basic_function/rl/indexer_replay/test_npu_return_indexer_topk_glm52.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+NUM_INDEXER_LAYERS = 6
+INDEX_TOPK = 2048
+
+
+class TestNPUGLM52ReturnIndexerTopk(CustomTestCase):
+    """End-to-end validation of GLM-5.2 NPU indexer-topk responses."""
+
+    @classmethod
+    def setUpClass(cls):
+        cls.process = popen_launch_server(
```
Comment (floatlibai) [inline test/registered/npu/basic_function/rl/indexer_replay/test_npu_return_indexer_topk_glm52.py:37]:
```sh
python3 -m sglang.launch_server \
        --model-path $MODEL_PATH \
        --attention-backend ascend \
        --device npu \
        --tp-size 4 \
        --trust-remote-code \
        --host 127.0.0.1 \
        --mem-fraction-static 0.84 \
        --port 8000 \
        --dp 4 \
        --disable-cuda-graph \
        --enable-dp-attention \
        --enable-dp-lm-head \
        --enable-return-indexer-topk \
        --json-model-override-args '{"num_hidden_layers": 6}'
```

File: python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py (hunk 2 above)
Comment (floatlibai) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py:891]: done

**Review by Ronald1995 — APPROVED** [review]
LGTM
